### PR TITLE
Added event wildcard matching

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #3250, #14978

This solution created in attempt to keep good performance.
In case no wildcard even specified no significant performance degradation will appear.

### CODE REVIEW

**Comment by samdark** [comment]
Love the approach and the code looks good at the first glance. Going to take a break and re-read it several times.

File: framework/base/Component.php [hunk 1]
```diff
@@ -456,6 +463,13 @@ public function behaviors()
     public function hasEventHandlers($name)
     {
         $this->ensureBehaviors();
+
+        foreach ($this->_eventWildcards as $wildcard => $handlers) {
+            if (StringHelper::matchWildcard($wildcard, $name) && !empty($handlers)) {
```
Comment (samdark) [inline framework/base/Component.php]: Condition could be reverted. empty check costs less.

File: framework/base/Event.php [hunk 2]
```diff
@@ -110,23 +137,40 @@ public static function on($class, $name, $handler, $data = null, $append = true)
     public static function off($class, $name, $handler = null)
     {
         $class = ltrim($class, '\\');
-        if (empty(self::$_events[$name][$class])) {
+        if (empty(self::$_events[$name][$class]) && empty(self::$_eventWildcards[$name][$class])) {
             return false;
         }
         if ($handler === null) {
             unset(self::$_events[$name][$class]);
```
Comment (samdark) [inline framework/base/Event.php]: Could be written as `unset(self::$_events[$name][$class], self::$_eventWildcards[$name][$class])`.

File: framework/base/Event.php [hunk 3]
```diff
@@ -153,9 +198,10 @@ public static function offAll()
      */
     public static function hasHandlers($class, $name)
     {
-        if (empty(self::$_events[$name])) {
+        if (empty(self::$_events[$name]) && empty(self::$_eventWildcards)) {
```
Comment (samdark) [inline framework/base/Event.php]: Conditions could be inverted. Checking single variable costs less then accessing array index.

File: framework/base/Event.php [hunk 4]
```diff
@@ -187,9 +251,20 @@ class_implements($class, true)
...
+        if (!empty(self::$_eventWildcards)) {
+            foreach (self::$_eventWildcards as $nameWildcard => $classHandlers) {
+                if (!StringHelper::matchWildcard($nameWildcard, $name)) {
+                    continue;
+                }
+                $wildcardEventHandlers = array_merge($wildcardEventHandlers, $classHandlers);
+            }
+        }
+
+        if (empty(self::$_events[$name]) && empty($wildcardEventHandlers)) {
```
Comment (samdark) [inline framework/base/Event.php]: Conditions could be inverted. Checking single variable costs less then accessing array index.

**Review by samdark — APPROVED** [review]
Overall looks great. Haven't found much things to improve.

File: framework/base/Component.php (hunk 1 above)
Comment (klimov-paul) [reply framework/base/Component.php]: optimized

File: framework/base/Event.php (hunk 3 above)
Comment (klimov-paul) [reply framework/base/Event.php]: optimized

File: framework/base/Event.php (hunk 4 above)
Comment (klimov-paul) [reply framework/base/Event.php]: optimized

File: framework/base/Event.php (hunk 2 above)
Comment (klimov-paul) [reply framework/base/Event.php]: I would better not do it: arguments are long, such refactoring will create a long statement, which is hard to comprehend.

**Review by SilverFire — APPROVED** [review]

**Review by dmirogin — APPROVED** [review]

**Review by sergeymakinen — APPROVED** [review]

**Comment by samdark** [comment]
That was good. Thank you, Paul and thanks everyone for reviews.

**Comment by macklus** [comment]
thanks on that ;-)

**Pull request merged**